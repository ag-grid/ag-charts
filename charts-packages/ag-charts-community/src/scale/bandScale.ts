import Scale from './scale';

/**
 * Maps a discrete domain to a continuous numeric range.
 * See https://github.com/d3/d3-scale#band-scales for more info.
 */
export class BandScale<D> implements Scale<D, number> {

    /**
     * Maps datum to its index in the {@link domain} array.
     * Used to check for duplicate datums (not allowed).
     */
    private index = new Map<D, number>();

    /**
     * The output range values for datum at each index.
     */
    private ordinalRange: number[] = [];

    /**
     * Contains unique datums only. Since `{}` is used in place of `Map`
     * for IE11 compatibility, the datums are converted `toString` before
     * the uniqueness check.
     */
    private _domain: D[] = [];
    set domain(values: D[]) {
        const domain = this._domain;
        domain.length = 0;

        this.index = new Map<D, number>();
        const index = this.index;

        // In case one wants to have duplicate domain values, for example, two 'Italy' categories,
        // one should use objects rather than strings for domain values like so:
        // { toString: () => 'Italy' }
        // { toString: () => 'Italy' }
        values.forEach(value => {
            if (index.get(value) === undefined) {
                index.set(value, domain.push(value) - 1);
            }
        });

        this.rescale();
    }
    get domain(): D[] {
        return this._domain;
    }

    private _range: [number, number] = [0, 1];
    set range(values: [number, number]) {
        this._range[0] = values[0];
        this._range[1] = values[1];
        this.rescale();
    }
    get range(): [number, number] {
        return this._range;
    }

    ticks(): D[] {
        return this._domain;
    }

    convert(d: D): number {
        const i = this.index.get(d);
        if (i === undefined) {
            return NaN;
        }

        const r = this.ordinalRange[i];
        if (r === undefined) {
            return NaN;
        }

        return r;
    }

    private _bandwidth: number = 1;
    get bandwidth(): number {
        return this._bandwidth;
    }

    set padding(value: number) {
        value = Math.max(0, Math.min(1, value));
        this._paddingInner = value;
        this._paddingOuter = value;
        this.rescale();
    }
    get padding(): number {
        return this._paddingInner;
    }

    /**
     * The ratio of the range that is reserved for space between bands.
     */
    private _paddingInner = 0;
    set paddingInner(value: number) {
        this._paddingInner = Math.max(0, Math.min(1, value)); // [0, 1]
        this.rescale();
    }
    get paddingInner(): number {
        return this._paddingInner;
    }

    /**
     * The ratio of the range that is reserved for space before the first
     * and after the last band.
     */
    private _paddingOuter = 0;
    set paddingOuter(value: number) {
        this._paddingOuter = Math.max(0, Math.min(1, value)); // [0, 1]
        this.rescale();
    }
    get paddingOuter(): number {
        return this._paddingOuter;
    }

    private _round = false;
    set round(value: boolean) {
        this._round = value;
        this.rescale();
    }
    get round(): boolean {
        return this._round;
    }

    /**
     * How the leftover range is distributed.
     * `0.5` - equal distribution of space before the first and after the last band,
     * with bands effectively centered within the range.
     */
    private _align = 0.5;
    set align(value: number) {
        this._align = Math.max(0, Math.min(1, value)); // [0, 1]
        this.rescale();
    }
    get align(): number {
        return this._align;
    }

    protected rescale() {
        const n = this._domain.length;
        if (!n) {
            return;
        }

        let [a, b] =  this._range;
        const reversed = b < a;

        if (reversed) {
            [a, b] = [b, a];
        }
        let step = (b - a) / Math.max(1, n - this._paddingInner + this._paddingOuter * 2);
        if (this._round) {
            step = Math.floor(step);
        }
        a += (b - a - step * (n - this._paddingInner)) * this._align;
        this._bandwidth = step * (1 - this._paddingInner);
        if (this._round) {
            a = Math.round(a);
            this._bandwidth = Math.round(this._bandwidth);
        }

        const values: number[] = [];
        for (let i = 0; i < n ; i++) {
            values.push(a + step * i);
        }

        this.ordinalRange = reversed ? values.reverse() : values;
    }
}
