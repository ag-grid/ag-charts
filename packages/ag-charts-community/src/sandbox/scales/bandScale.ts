import { BaseScale } from './baseScale';

type DomainType = Date | number | string;

export interface BandScaleOptions {
    align?: number;
    paddingInner?: number;
    paddingOuter?: number;
    round?: boolean;
}

export class BandScale extends BaseScale<DomainType, number> {
    private readonly ordinalRange: number[];

    readonly align: number;
    readonly paddingInner: number;
    readonly paddingOuter: number;
    readonly round: boolean;

    constructor(domain: DomainType[], range: number[], options?: BandScaleOptions) {
        super(domain, range);

        this.align = options?.align ?? 0.5;
        this.paddingInner = options?.paddingInner ?? 0.1;
        this.paddingOuter = options?.paddingOuter ?? 0.1;
        this.round = options?.round ?? false;

        const [r0, r1] = range;
        const step = this.step();

        let start = Math.min(r0, r1) + (Math.abs(r1 - r0) - step * (domain.length - this.paddingInner)) * this.align;

        if (this.round) {
            start = Math.round(start);
        }

        this.ordinalRange = domain.map((_, i) => start + step * i);
    }

    convert(value: DomainType) {
        return this.ordinalRange[this.domain.indexOf(value)];
    }

    invert(value: number) {
        let matchIndex = -1;
        for (const [position, index] of this.ordinalRange.entries()) {
            if (position > value) break;
            matchIndex = index;
        }
        return this.domain[matchIndex];
    }

    bandwidth() {
        const bandwidth = this.step() * (1 - this.paddingInner);
        return this.round ? Math.round(bandwidth) : bandwidth;
    }

    step() {
        const [r0, r1] = this.range;
        const count = this.domain.length;
        const { paddingInner, paddingOuter } = this;
        const step = Math.abs(r1 - r0) / Math.max(1, count - paddingInner + paddingOuter * 2);
        return this.round ? Math.floor(step) : step;
    }

    override reverse() {
        super.reverse();
        this.ordinalRange.reverse();
    }
}
