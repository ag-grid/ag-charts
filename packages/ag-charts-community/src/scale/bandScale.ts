import { Logger } from '../util/logger';
import { clamp } from '../util/number';
import { Invalidating } from './invalidating';
import type { Scale } from './scale';

/**
 * Maps a discrete domain to a continuous numeric range.
 */
export class BandScale<D, I = number> implements Scale<D, number, I> {
    static is(value: any): value is BandScale<any, any> {
        return value instanceof BandScale;
    }

    readonly type: 'band' | 'ordinal-time' = 'band';

    protected invalid = true;

    @Invalidating
    range: number[] = [0, 1];

    @Invalidating
    round = false;

    @Invalidating
    interval?: I = undefined;

    protected refresh() {
        if (!this.invalid) return;

        this.invalid = false;
        this.update();

        if (this.invalid) {
            Logger.warnOnce('Expected update to not invalidate scale');
        }
    }

    /**
     * Maps datum to its index in the {@link domain} array.
     * Used to check for duplicate data (not allowed).
     */
    protected index = new Map<D | D[], number>();

    /**
     * The output range values for datum at each index.
     */
    protected ordinalRange: number[] = [];

    /**
     * Contains unique data only.
     */
    protected _domain: D[] = [];
    set domain(values: D[]) {
        this.index = new Map<D, number>();
        this.invalid = true;
        this._domain = [];

        // In case one wants to have duplicate domain values, for example, two 'Italy' categories,
        // one should use objects rather than strings for domain values like so:
        // { toString: () => 'Italy' }
        // { toString: () => 'Italy' }
        for (const value of values) {
            const key = value instanceof Date ? (value.getTime() as D) : value;
            if (this.getIndex(key) === undefined) {
                this.index.set(key, this._domain.push(value) - 1);
            }
        }
    }
    get domain(): D[] {
        return this._domain;
    }

    ticks(): { ticks: D[]; fractionDigits: number } {
        this.refresh();
        return { ticks: this._domain, fractionDigits: 0 };
    }

    convert(d: D): number {
        this.refresh();
        const i = this.getIndex(d);
        if (i == null) {
            return NaN;
        }
        return this.ordinalRange[i] ?? NaN;
    }

    invert(position: number) {
        this.refresh();
        const index = this.ordinalRange.findIndex((p) => p === position);
        return this.domain[index];
    }

    invertNearest(position: number) {
        this.refresh();
        let nearest = -1;
        let minDistance = Infinity;

        const index = this.ordinalRange.findIndex((p, i) => {
            if (p === position) return true;
            const distance = Math.abs(position - p);
            if (distance < minDistance) {
                minDistance = distance;
                nearest = i;
            }
            return false;
        });

        return this.domain[index] ?? this.domain[nearest];
    }

    private _bandwidth: number = 1;
    get bandwidth(): number {
        this.refresh();
        return this._bandwidth;
    }

    private _step: number = 1;
    get step(): number {
        this.refresh();
        return this._step;
    }

    private _rawBandwidth: number = 1;
    get rawBandwidth(): number {
        this.refresh();
        return this._rawBandwidth;
    }

    set padding(value: number) {
        value = clamp(0, value, 1);
        this._paddingInner = value;
        this._paddingOuter = value;
    }
    get padding(): number {
        return this._paddingInner;
    }

    /**
     * The ratio of the range that is reserved for space between bands.
     */
    private _paddingInner = 0;
    set paddingInner(value: number) {
        this._paddingInner = clamp(0, value, 1);
    }
    get paddingInner(): number {
        return this._paddingInner;
    }

    /**
     * The ratio of the range that is reserved for space before the first
     * and after the last band.
     */
    private _paddingOuter = 0;
    set paddingOuter(value: number) {
        this._paddingOuter = clamp(0, value, 1);
    }
    get paddingOuter(): number {
        return this._paddingOuter;
    }

    update() {
        const count = this._domain.length;

        if (count === 0) return;

        const [r0, r1] = this.range;
        let { _paddingInner: paddingInner } = this;
        const { _paddingOuter: paddingOuter, round } = this;
        const rangeDistance = r1 - r0;

        let rawStep: number, step: number, start: number;

        if (count === 1) {
            paddingInner = 0;
            rawStep = rangeDistance * (1 - paddingOuter * 2);
            step = round ? Math.round(rawStep) : rawStep;
            start = rangeDistance * paddingOuter;
        } else {
            rawStep = rangeDistance / Math.max(1, count - paddingInner + paddingOuter * 2);
            step = round ? Math.floor(rawStep) : rawStep;
            start = r0 + (rangeDistance - step * (count - paddingInner)) / 2;
        }

        let bandwidth = step * (1 - paddingInner);

        if (round) {
            start = Math.round(start);
            bandwidth = Math.round(bandwidth);
        }

        this._step = step;
        this._bandwidth = bandwidth;
        this._rawBandwidth = rawStep * (1 - paddingInner);
        this.ordinalRange = this._domain.map((_, i) => start + step * i);
    }

    private getIndex(value: D) {
        return this.index.get(value instanceof Date ? (value.getTime() as D) : value);
    }
}
