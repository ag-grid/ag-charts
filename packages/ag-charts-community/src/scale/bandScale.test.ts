import { describe, expect, it, test } from '@jest/globals';

import { BandScale } from './bandScale';

describe('BandScale', () => {
    test('initial state', () => {
        const scale = new BandScale();

        expect(scale.domain).toEqual([]);
        expect(scale.range).toEqual([0, 1]);
        expect(scale.paddingInner).toBe(0);
        expect(scale.paddingOuter).toBe(0);
        expect(scale.padding).toBe(0);
    });

    test('no implicit domain construction', () => {
        const scale = new BandScale();

        expect(scale.convert('B')).toBe(NaN);
        expect(scale.convert('C')).toBe(NaN);
        expect(scale.convert('A')).toBe(NaN);

        expect(scale.domain).toEqual([]);
    });

    test('basic band scale', () => {
        const scale = new BandScale();

        scale.domain = ['A', 'B', 'C', 'D', 'E'];
        scale.range = [0, 500];

        expect(scale.convert('A')).toBe(0);
        expect(scale.convert('B')).toBe(100);
        expect(scale.convert('C')).toBe(200);
        expect(scale.convert('D')).toBe(300);
        expect(scale.convert('E')).toBe(400);

        expect(scale.bandwidth).toBe(100);
    });

    test('date band scale', () => {
        const scale = new BandScale();

        scale.domain = [new Date('2020'), new Date('2021'), new Date('2022'), new Date('2023'), new Date('2024')];
        scale.range = [0, 500];

        expect(scale.convert(new Date('2020'))).toBe(0);
        expect(scale.convert(new Date('2021'))).toBe(100);
        expect(scale.convert(new Date('2022'))).toBe(200);
        expect(scale.convert(new Date('2023'))).toBe(300);
        expect(scale.convert(new Date('2024'))).toBe(400);

        expect(scale.bandwidth).toBe(100);
    });

    test('inner padding', () => {
        const scale = new BandScale();

        scale.domain = ['A', 'B', 'C', 'D', 'E'];
        scale.range = [0, 500];
        scale.paddingInner = 1;

        expect(scale.convert('A')).toBe(0);
        expect(scale.convert('B')).toBe(125);
        expect(scale.convert('C')).toBe(250);
        expect(scale.convert('D')).toBe(375);
        expect(scale.convert('E')).toBe(500);
    });

    test('inner padding with single value', () => {
        const scale = new BandScale();

        scale.domain = ['A'];
        scale.range = [0, 500];
        scale.paddingInner = 0.5;

        expect(scale.convert('A')).toBe(0);
    });

    test('outer padding', () => {
        const scale = new BandScale();

        scale.domain = ['A', 'B', 'C', 'D', 'E'];
        scale.range = [0, 500];
        scale.paddingOuter = 0.06;

        expect(scale.convert('A')).toBe(5.859375);
        expect(scale.convert('B')).toBe(103.515625);
        expect(scale.convert('C')).toBe(201.171875);
        expect(scale.convert('D')).toBe(298.828125);
        expect(scale.convert('E')).toBe(396.484375);
    });

    test('padding', () => {
        const scale = new BandScale();

        scale.domain = ['A', 'B', 'C', 'D', 'E'];
        scale.range = [0, 500];
        scale.paddingInner = 1;
        scale.paddingOuter = 0.5;

        expect(scale.convert('A')).toBe(50);
        expect(scale.convert('B')).toBe(150);
        expect(scale.convert('C')).toBe(250);
        expect(scale.convert('D')).toBe(350);
        expect(scale.convert('E')).toBe(450);

        expect(scale.bandwidth).toBe(0);
    });

    test('round', () => {
        const scale = new BandScale();

        scale.domain = ['A', 'B', 'C', 'D', 'E'];
        scale.range = [0, 500];
        scale.paddingOuter = 0.06;
        scale.round = true;

        expect(scale.convert('A')).toBe(8);
        expect(scale.convert('B')).toBe(105);
        expect(scale.convert('C')).toBe(202);
        expect(scale.convert('D')).toBe(299);
        expect(scale.convert('E')).toBe(396);

        expect(scale.bandwidth).toBe(97);
    });

    describe('should create ticks', () => {
        const CASES = [
            {
                interval: 0,
                domain: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
            },
            {
                interval: 1,
                domain: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
            },
            {
                interval: 2,
                domain: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
            },
            {
                interval: 3,
                domain: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
            },
            {
                interval: 4,
                domain: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
            },
            {
                interval: -1,
                domain: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
            },
            {
                interval: -2,
                domain: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
            },
            {
                interval: 1.5,
                domain: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
            },
            {
                interval: -1.5,
                domain: ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
            },
        ];

        it.each(CASES)(`for interval: $interval domain: $domain case`, ({ interval, domain }) => {
            const scale = new BandScale();

            scale.range = [0, 600];
            scale.domain = domain;
            scale.interval = interval;

            expect(scale.ticks()).toMatchSnapshot();
        });
    });
});
