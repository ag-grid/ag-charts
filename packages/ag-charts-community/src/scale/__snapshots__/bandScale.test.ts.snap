// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create ticks for interval: -1 domain: [A, B] case 1`] = `
Array [
  "A",
  "B",
  "C",
  "D",
  "E",
  "F",
  "G",
  "H",
  "I",
]
`;

exports[`should create ticks for interval: -1.5 domain: [A, B] case 1`] = `
Array [
  "A",
  "B",
  "C",
  "D",
  "E",
  "F",
  "G",
  "H",
  "I",
]
`;

exports[`should create ticks for interval: -2 domain: [A, B] case 1`] = `
Array [
  "A",
  "C",
  "E",
  "G",
  "I",
]
`;

exports[`should create ticks for interval: 0 domain: [A, B] case 1`] = `Array []`;

exports[`should create ticks for interval: 1 domain: [A, B] case 1`] = `
Array [
  "A",
  "B",
  "C",
  "D",
  "E",
  "F",
  "G",
  "H",
  "I",
]
`;

exports[`should create ticks for interval: 1.5 domain: [A, B] case 1`] = `
Array [
  "A",
  "C",
  "E",
  "G",
  "I",
]
`;

exports[`should create ticks for interval: 2 domain: [A, B] case 1`] = `
Array [
  "A",
  "C",
  "E",
  "G",
  "I",
]
`;

exports[`should create ticks for interval: 3 domain: [A, B] case 1`] = `
Array [
  "A",
  "D",
  "G",
]
`;

exports[`should create ticks for interval: 4 domain: [A, B] case 1`] = `
Array [
  "A",
  "E",
  "I",
]
`;
